src/dvclive/monitor_system.py

```diff
@@ -0,0 +1,265 @@
+
+    Raises:
+        ValueError: if the keys in `directories_to_monitor` contains invalid characters
+        as defined by `os.path.normpath`.
+    """
+
+    _plot_blacklist_prefix: Tuple = (
+        METRIC_CPU_COUNT,
+        METRIC_RAM_TOTAL_GB,
+        METRIC_DISK_TOTAL_GB,
+        METRIC_GPU_COUNT,
+        METRIC_VRAM_TOTAL_GB,
+    )
+
+    def __init__(
+        self,
+        interval: float = 0.05,  # seconds
+        num_samples: int = 20,
+        directories_to_monitor: Optional[Dict[str, str]] = None,
+        plot: bool = True,
+    ):
+        self._interval = self._check_interval(interval, max_interval=0.1)
+        self._num_samples = self._check_num_samples(
+            num_samples, min_num_samples=1, max_num_samples=30
+        )
+        self._disks_to_monitor = self._check_directories_to_monitor(
+            directories_to_monitor
+        )
+        self._plot = plot
+        self._warn_cpu_problem = True
+        self._warn_gpu_problem = True
+        self._warn_disk_doesnt_exist: Dict[str, bool] = {}
+
+    def _check_interval(self, interval: float, max_interval: float) -> float:
+        if interval > max_interval:
+            logger.warning(
+                f"System monitoring `interval` should be less than {max_interval} "
+                f"seconds. Setting `interval` to {max_interval} seconds."
+            )
+            return max_interval
+        return interval
+
+    def _check_num_samples(
+        self, num_samples: int, min_num_samples: int, max_num_samples: int
+    ) -> int:
+        min_num_samples = 1
+        max_num_samples = 30
+        if not min_num_samples < num_samples < max_num_samples:
+            num_samples = max(min(num_samples, max_num_samples), min_num_samples)
+            logger.warning(
+                f"System monitoring `num_samples` should be between {min_num_samples} "
+                f"and {max_num_samples}. Setting `num_samples` to {num_samples}."
+            )
+        return num_samples
+
+    def _check_directories_to_monitor(
+        self, directories_to_monitor: Optional[Dict[str, str]]
+    ) -> Dict[str, str]:
+        if directories_to_monitor is None:
+            return {"main": "/"}
+
+        disks_to_monitor = {}
+        for disk_name, disk_path in directories_to_monitor.items():
+            if disk_name != os.path.normpath(disk_name):
+                raise ValueError(  # noqa: TRY003
+                    "Keys for `directories_to_monitor` should be a valid name"
+                    f", but got '{disk_name}'."
+                )
+            disks_to_monitor[disk_name] = disk_path
+        return disks_to_monitor
+
+    def __call__(self, live):
+        self._live = live
+        self._shutdown_event = Event()
+        Thread(
+            target=self._monitoring_loop,
+        ).start()
+
+    def _monitoring_loop(self):
+        while not self._shutdown_event.is_set():
```

Comment:
Can we put the rest of this logic into its own method? That way, it should be easier to test the underlying logic without having to create a thread that has to be shut down.

Reply:
I'm not quite sure I understand what you mean here :) 
There is lots of logic with the _shutdown_event inside the following lines, how should I handle them?

Reply:
Okay, sorry, didn't read/think through it carefully. If the tests are all stable now, no concerns from me.